lib/vkloader.c

```diff
@@ -1763,18 +1873,62 @@ generateMipmaps(ktxVulkanTexture* vkTexture, ktxVulkanDeviceInfo* vdi,
  *
  * Frees the Vulkan resources created when the texture image was loaded.
+ * If a complete set of suballocator callbacks
+ * are provided, it will use the free call from those instead.
  *
- * @param vkTexture  pointer to the ktxVulkanTexture to be destructed.
- * @param device     handle to the Vulkan logical device to which the texture was
- *                   loaded.
- * @param pAllocator pointer to the allocator used during loading.
+ * @param vkTexture                pointer to the ktxVulkanTexture to be destructed.
+ * @param device                   handle to the Vulkan logical device to which the texture was
+ *                                 loaded.
+ * @param pAllocator               pointer to the allocator used during loading.
+ * @param subAllocatorCallbacks    potential function pointer for guarded suballocator callbacks.
+ *                                 Pass NULL if a suballocator was not used.
+ * @return                         KTX_SUCCESS on success, KTX_INVALID_OPERATION if the
+ *                                 supplied subAllocatorCallbacks structure is incomplete.
+ */
+ktx_error_code_e
+ktxVulkanTexture_Destruct_WithSuballocator(ktxVulkanTexture* vkTexture, VkDevice device,
+                                           const VkAllocationCallbacks* pAllocator,
+                                           ktxVulkanTexture_subAllocatorCallbacks* subAllocatorCallbacks)
+{
+    ktx_bool_t useSuballocator = false;
+    if (subAllocatorCallbacks) {
+        if (subAllocatorCallbacks->allocMemFuncPtr &&
+            subAllocatorCallbacks->bindBufferFuncPtr &&
+            subAllocatorCallbacks->bindImageFuncPtr &&
+            subAllocatorCallbacks->memoryMapFuncPtr &&
+            subAllocatorCallbacks->memoryUnmapFuncPtr &&
+            subAllocatorCallbacks->freeMemFuncPtr)
+            useSuballocator = true;
+        else
+            return KTX_INVALID_OPERATION;
+    }
+
+    vkTexture->vkDestroyImage(device, vkTexture->image, pAllocator);
+    if (useSuballocator)
+        subAllocatorCallbacks->freeMemFuncPtr(vkTexture->allocationId);
+    else
+        vkTexture->vkFreeMemory(device, vkTexture->deviceMemory, pAllocator);
+
+    return KTX_SUCCESS;
+}
+
+/**
+ * @memberof ktxVulkanTexture
+ * @~English
+ * @brief Destructor for the object returned when loading a texture image.
+ *
+ * Calls ktxVulkanTexture_Destruct_WithSuballocator() without a
+ * set of suballocator callbacks.
+ *
+ * @sa ktxVulkanTexture_Destruct_WithSuballocator() for details and use that for complete
+ *     control.
  */
 void
 ktxVulkanTexture_Destruct(ktxVulkanTexture* vkTexture, VkDevice device,
                           const VkAllocationCallbacks* pAllocator)
 {
-    vkTexture->vkDestroyImage(device, vkTexture->image, pAllocator);
-    vkTexture->vkFreeMemory(device, vkTexture->deviceMemory, pAllocator);
+    ktxVulkanTexture_Destruct_WithSuballocator(vkTexture, device, pAllocator, NULL);
```

Comment:
s/ktxVulkan/(void)ktxVulkan/ to indicate we know we are ignoring the return value.

Reply:
👍